.github/workflows/main.yml

```diff
@@ -168,7 +168,7 @@ jobs:
         shell: Rscript {0}
 
       - name: Check
-        run: Rscript -e "rcmdcheck::rcmdcheck('${{ needs.jobR.outputs.r_bindings }}', args = c('--no-manual','--as-cran'), error_on = 'warning', check_dir = 'check')"
+        run: Rscript -e "rcmdcheck::rcmdcheck('${{ needs.jobR.outputs.r_bindings }}', args = c('--no-manual','--as-cran'), error_on = 'error', check_dir = 'check')"
```

Comment:
Since CRAN will reject for warnings, should we revert this to `warning`?  You know better than me.

Reply:
Yes. We should revert _once we have a new RcppEnsmallen at CRAN_.  Until then we are guaranteed

- a warning because of the no-warning-given deprecatation in Armadillo 14.2.0
- a guaranteed fail when we error on warning

which is why I flipped this for now.  You as maintainer are free to overrule but you then have to live with all the :x: that is entails :scream_cat:

Reply:
> Since CRAN will reject for warnings

My approach with my (smaller !!) CRAN packages has been to deal with that locally not force my CIs to mail over what are often transient nuisances.  But different strokes for different folks -- you are clearly a much a much more patient person than I am.  I dislike it when CI runs take over ten minutes.  Several hours, spread over different backend systems is ... a different cup of tea.

Reply:
Yeah unfortunately with how long mlpack takes to compile in general (as with any heavily template metaprogrammed C++ application...), I don't have any way to get away from long turnaround times for CI.  So I'm not a very patient person in general but I have had to learn some of it for our CI infrastructure.  (I guess that's probably a good thing for me as a person overall.)

I saw your comment about releasing a new version of ensmallen---I think that's a good idea and I'll probably do it over the weekend, just a bugfix patch or something.  That in turn should mean a fairly quick turnaround for a new version of RcppEnsmallen, and then we have no issue here.  Personally I would rather do that than disable the error and forget to revert (which honestly I think is fairly high probability).

> no-heads-up-given-to-us

Technically the whole thing is my fault :smile: I noticed that the overload was not documented and pointed it out.  And honestly I think a deprecation warning is a totally reasonable way to provide the heads-up (in fact that's what deprecation is: a heads-up).  My issue is more with CRAN's policy that makes the deprecation warning suddenly unacceptable.  (Of course, the good side of that: the deprecated function usage gets resolved very quickly!)

Reply:
sidenote: I've empirically found that using `ionice -c 3` in conjunction with standard `nice` tends to make multi-core compilation of mlpack less taxing; example:

    $ ionice -c 3 nice make -j 4 mlpack_test

`ionice -c 3` prevents I/O access by a process until the I/O queue is clear: https://booksonline.nl/man/htmlman1/ionice.1.html

Reply:
Re 'no heads-up given' and 'technically my fault':  I am for deprecation too, but you MUST managed them downstream if someone like CRAN checks.  More than once I had a really long open issue tackling every downstream package til a patch was in _at CRAN_. See eg [this one](https://github.com/RcppCore/RcppArmadillo/issues/391) for an example.  So in my view the only way we can handle this is to 
- create a deprecation warning as opt-in, leave it off by default
- turn it, find all affected downstream packages, changes all of them (!!)
- only once that is done, turn it on by default

Consequently, RcppArmadillo is forced to turn this off, as upsetting as that is for @conradsnicta.  It affects too many packages.  `mlpack` got bitten at CRAN because it uses Arma header directly.

Re `ionice`:  Awesome. Will try.  I had to give up on `make -j 4` so can try this.

`ccache` also helps as it reduces what gets compiled.  I may open an issue.

Reply:
@rcurtin I can of course also revert this change now, given that we get a merge over :x: anyway.  Let me know if you'd prefer that.

And thanks for the review @conradsnicta !!

Reply:
@rcurtin I will leave this to you and revert it now as it is not really a core part of the PR, and did not help in "getting to :heavy_check_mark:" which other parts failing.